packages/tpetra/core/src/Tpetra_MultiVector_def.hpp

```diff
@@ -2898,6 +2898,60 @@ void MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>::copyAndPermute(
   }
 
 
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void
+  MultiVector<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  scale (const execution_space& exec, const Scalar& alpha)
+  {
+    using Kokkos::ALL;
+    using IST = impl_scalar_type;
+
+    const IST theAlpha = static_cast<IST> (alpha);
+    if (theAlpha == Kokkos::ArithTraits<IST>::one ()) {
+      return; // do nothing
+    }
+    const size_t lclNumRows = getLocalLength ();
```

Comment:
Why would you make these copy-pastes ? I think we should really put some effort in designing this carefully so we avoid code duplication, which is highway to hell :smiling_imp:

Reply:
To be honest... I don't really like it much. It would make the code hard to read, and a pain to maintain.

In `Kokkos`, what they usually do (@masterleinad Could you confirm this is still the way to go ?) is that the overload that does not take an execution space instance first globally fences, then uses the other overload with the default execution space instance, and then fence. Like this:
```c++
template <typename ViewType>
void my_func(const ViewType& data)
{
    Kokkos::fence(...);
    typename ViewType::execution_space exec;
    my_func(exec, data);
    exec.fence(...);
}

template <typename Exec, typename ViewType>
void my_func(const Exec& exec, const ViewType& data)
{
    ...
}
```

I think this is a good design:
- the implementation is still easy to read, and to maintain because it will always work with an execution space instance (no need for `constexpr if` statements that kill the readability)
- for a user that does not care much about asynchronicity, I think it's no big deal to have a global fence from time to time.

Here is an example: https://github.com/kokkos/kokkos/blob/33c9b8cef70eddabc3c1f07d053be23abf81e912/algorithms/src/Kokkos_Random.hpp#L1554-L1563.

Reply:
@csiefer2 I'll be out of office for 2 weeks. Would you like to discuss it in a dedicated meeting somewhere around the 25th of July?

Reply:
Did this get sorted out?

Reply:
Sorry, this went out of my head, but we're definitely interested in discussing this !